Ursa port

GeorgeGayno-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 

- Port repository to Ursa. 
- Remove support for Hera, S4 and C5.
- Updated the Doxygen config files to work with Ursa's newer version of the Doxygen executable.

## TESTS CONDUCTED: 

- [x] Compile branch on all Tier 1 machines using Intel (Orion, Jet, Ursa, Hercules and WCOSS2). Done using c254bd8.
- [x] Compile branch on Ursa using GNU. Done using 5e0310b.
- [x] Compile branch in 'Debug' mode on WCOSS2. Done using 5e0310b.
- [x] Compile with Doxygen on any machine with no errors. Done on Ursa and Dogwood using f416e22
- [x] Run unit tests locally on any Tier 1 machine. Done on Ursa using Gnu and 5e0310b. All tests passed.
- [x] Run all consistency tests locally on Ursa. The baselines on Hera were moved to Ursa for these tests. **See below for results.**

Consistency test results:
- `cpld_gridgen` - https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996546983 and https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-3019040777
- `chgres_cube` - https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996526969
- `grid_gen` - https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996509274
- `ocnice_prep` - https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996385394 and https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-3013424664
- `regrid_sfc` - https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996311409
- `global_cycle`, `weight_gen`, `snow2mdl`, `ice_blend` - https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996272393

Additional tests performed:

- A C96 global uniform grid was successfully created using `./driver_scripts/driver_grid.ursa.sh` (f416e22)
- The `weight_gen` utility (`./util/weight_gen`) was successfully run. (f416e22)
- The `sfc_climo_gen` utility (`./util/sfc_climo_gen`) was successfully run on a C768 global grid. (f416e22)
- The `gdas_init` utility (`./util/gdas_init`) was successfully run for the following cases:
  - 20240101/06z (OPS v16 data as input); GDAS cycle; C192/C96; (f416e22)
  - 20240101/06z (OPS v16 data as input); GFS cycle; C768; (f416e22)
  - 20210101/06z (OPS v15 data as input); GDAS cycle; C192/C96; (f416e22)
  - 20210101/06z (OPS v15 data as input); GFS cycle; C768; (f416e22)
  - 20210101/06z (v16 retrospective data as input); GDAS cycle; C768/C384; (f416e22)
  - 20210101/06z (v16 retrospective data as input); GFS cycle; C768; (f416e22)
 -  All tools in `./sorc/orog_mask_tools.fd/orog_netcdf_tools.fd` were successfully run. (f416e22)

## DEPENDENCIES:
- ufs-community/ufs-weather-model#2471

## DOCUMENTATION:
The RST file was updated (references to Hera were removed and to Ursa were added): [See here](https://georgegayno-noaaufs-utils.readthedocs.io/en/ursa_port/ufs_utils.html#ufs-utils-utilities).

## ISSUE: 
Fixes #1064.

GeorgeGayno-NOAA committed:
Add build module for ursa.

Update source code to access required modules for ip v5.

Fixes #1064.

GeorgeGayno-NOAA committed:
Remove references to Hera from PR template.

Fixes #1064.

GeorgeGayno-NOAA committed:
Add pre-processing directives so code can compile with
pre-V5 versions of IPLIB.

Fixes #1064.

GeorgeGayno-NOAA committed:
Port weight_gen regression test.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update regrid_sfc regression test driver script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update ./fix/link_fixdirs.sh for ursa. Remove references
to hera.

Fixes #1064.

GeorgeGayno-NOAA committed:
Port ice blend regression test script to ursa.

Fixes #1064.

GeorgeGayno-NOAA committed:
Port snow2mdl regression test.

Fixes #1064.

GeorgeGayno-NOAA committed:
Point to official baseline directories.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update ocnice_prep regression test script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update cpld_gridgen regression test script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update grid_gen regression test script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update the chgres_cube regression test.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update global_cycle regression test.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update weight_gen utility script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Remove hera from ./sorc/machine-setup.sh.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update the datmmesh_gen and sfc_climo_gen utilities.

Fixes #1064.

GeorgeGayno-NOAA committed:
Begin updates to gdas_init utility.

Fixes #1064.

GeorgeGayno-NOAA committed:
Finish modifications to gdas_init.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update README.md.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update local unit tests.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update grid driver script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Remove hera build modules. Add ursa build module using gnu.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update comments in CMakeLists.txt and build_all.sh.

Fixes #1064.

GeorgeGayno-NOAA committed:
Convert the orog utilities to run on ursa.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update 'readthedocs'.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update README.md.

Fixes #1064.

GeorgeGayno-NOAA committed:
Some final updates.

Fixes #1064.

GeorgeGayno-NOAA committed:
Remove 's4' support from gdas_init utility.

Fixes #1064.

GeorgeGayno-NOAA committed:
Remove more 's4' support.

Fixes #1064.

GeorgeGayno-NOAA committed:
Remove references to GaeaC5.

Fixes #1064.

GeorgeGayno-NOAA committed:
Load ip v5 AND sp so I don't need to modify sfcsub.F.

Fixes #1064.

GeorgeGayno-NOAA committed:
Merge branch 'develop' into ursa_port

Fixes #1064.

BrianCurtis-NOAA commented on CMakeLists.txt:
```diff
@@ -121,9 +119,12 @@ if(CHGRES_ALL OR GBLEVENTS)
 find_package(ip 3.3.3 REQUIRED)
-if(ip_VERSION LESS 5.0)
+# Can't use ip v5 without sp because modfications would need to be
+# made to global_cycle routine sfcsub.F, which is part of
+# ccpp-physics repo.
+#if(ip_VERSION LESS 5.0)
   find_package(sp 2.3.3 REQUIRED)
-endif()
+#endif()
```
Could this be shortened to:
```
# Please don't use ip v5 without sp
```
and remove lines 123-125 and 127 ?

DavidHuber-NOAA replied:
I think at least a `TODO` should remain.  We should be migrating to ip for all sp functionality as sp will no longer be maintained.  This is just a postponement until the ccpp-physics package has the proper `#ifdef` blocks added.  Perhaps:

```suggestion
# TODO: Disable sp when using ip v5.0+.
#       This requires an update to ccpp-physics and global_cycle's sfcsub.F module.
  find_package(sp 2.3.3 REQUIRED)
#endif()
```

GeorgeGayno-NOAA replied:
@DavidHuber-NOAA - agree. Will use your wording.

GeorgeGayno-NOAA replied:
Done at 84f78f7.

BrianCurtis-NOAA commented on cmake/mpiexec.ursa:
```diff
@@ -12,4 +12,4 @@ QOS=debug
 shift
 
-srun -A $ACCOUNT -q $QOS -n $NP $@
+srun -A $ACCOUNT -q $QOS -t 0:01 -n $NP $@
```
confirming, no more than 1 minute for all items that use this?

GeorgeGayno-NOAA replied:
Yes. Each individual test runs in under 10 seconds. Most run in under 1 second.  (There are 55 tests). If you don't specify the wall clock time, you are given a default of 5 minutes. One day I was getting slow turnaround from the queue. So, I set it to 1 minute. I don't know if it made a difference or not.

BrianCurtis-NOAA commented on modulefiles/build.ursa.gnu.lua:
```diff
@@ -0,0 +1,60 @@
+Load environment to compile UFS_UTILS on Ursa using Gnu
+]])
+
+prepend_path("MODULEPATH", "/contrib/spack-stack/spack-stack-1.9.1/envs/ue-gcc-12.4.0/install/modulefiles/Core")
```
Noting, this might be 1.9.2 by the time the UFSWM Ursa port is completed.

GeorgeGayno-NOAA replied:
Should we hold off on merging until 1.9.2 is ready?

BrianCurtis-NOAA replied:
The UFSWM timeline is spack-stack 1.9.1 PR then Ursa port PR. I think 1.9.2 will be installed within the week on machines, but i don't know what they wanted to do with the UFSWM SS-191 PR.

BrianCurtis-NOAA replied:
@GeorgeGayno-NOAA looks like they've made 1.9.2 the version of spack-stack for the ufswm and ursa PR's. Here is the Ursa PR that you might want to use similar locations for this PR: https://github.com/ufs-community/ufs-weather-model/pull/2471/files#diff-2cf9f92d4e786470bafa8ca7b6311ecb2279ecdf03a3433abe099a3100c6a55d

BrianCurtis-NOAA replied:
Sorry, this is the GNU version: https://github.com/ufs-community/ufs-weather-model/pull/2471/files#diff-033fcac8205710bdd276f468829a13bcc0cf0b61569f76734069c1091e48d1ca

GeorgeGayno-NOAA replied:
Done at c254bd8.

BrianCurtis-NOAA commented on modulefiles/build.ursa.intelllvm.lua:
```diff
@@ -1,16 +1,16 @@
 ]])
 
-prepend_path("MODULEPATH", "/data/prod/jedi/spack-stack/spack-stack-1.6.0/envs/upp-addon-env/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/modulefiles/Core")
```
Same note on 1.9.2

BrianCurtis-NOAA replied:
here's the Ursa intelllvm going into UFSWM: https://github.com/ufs-community/ufs-weather-model/pull/2471/files#diff-b1cba59a13bc0b3b43a5a65eea775f08e42552877640dd80e2ded02dd00cea0a

GeorgeGayno-NOAA replied:
Done at c254bd8.

BrianCurtis-NOAA commented on reg_tests/rt.sh:
```diff
@@ -43,8 +43,8 @@ else
-  elif [[ -d /scratch1 ]] ; then
-    target=Hera
+  elif [[ -d /scratch3 ]] ; then
+    target=Usra
```
Typo Usra -> Ursa

GeorgeGayno-NOAA replied:
Thanks. Fixed at d98c868.

BrianCurtis-NOAA commented on reg_tests/rt.sh:
```diff
@@ -88,8 +88,8 @@ cd ../reg_tests
 
-#if [[ $target == "orion" ]] || [[ $target == "jet" ]] || [[ $target == "hera" ]] || [[ $target == "hercules" ]] || [[ $target == "wcoss2" ]] ; then
-#if [[ $target == "orion" ]] || [[ $target == "jet" ]] || [[ $target == "hera" ]] || [[ $target == "hercules" ]] ; then
+#if [[ $target == "orion" ]] || [[ $target == "jet" ]] || [[ $target == "hercules" ]] || [[ $target == "wcoss2" ]] ; then
+#if [[ $target == "orion" ]] || [[ $target == "jet" ]] || [[ $target == "hercules" ]] ; then
```
Do we need to keep if they're going to stay commented out?

GeorgeGayno-NOAA replied:
Cleaned up script at ecfced3. FYI. This is the script that is started off the cron on each machine (using the role account)

BrianCurtis-NOAA reviewed (changes requested):
Need to fix a typo, the rest are notes for later.

GeorgeGayno-NOAA commented:
All `ocnice_prep` tests failed on Ursa (6bceac5). Differences from the (Hera) baseline appear to be small. Here are the differences from the `nccmp_050_ocean.log` file:

```
Variable Group    Count         Sum    AbsSum         Min        Max      Range         Mean      StdDev
sfc      /       219941 -0.00608705 0.0396815 -0.00443655 0.00203139 0.00646795 -2.76758e-08 1.65466e-05
eta      /     11540193     -530683    616446    -1902.77    414.788    2317.56   -0.0459856     5.24544
Temp     /     13415140    -1167.99   1793.17    -6.70048    2.81754    9.51802 -8.70649e-05   0.0212207
Salt     /      8242225     98.0152   284.894   -0.440129    1.37529    1.81541  1.18918e-05    0.004976
h        /     10253080     17246.3     20784    -252.123    411.662    663.786   0.00168206    0.586818
u        /     19370198    -1.17202   51.7173  -0.0749373  0.0754253   0.150363 -6.05062e-08 0.000273376
v        /     19330990     6.62261   46.9424   -0.037016  0.0818302   0.118846  3.42591e-07 0.000240951
```

@DeniseWorthen - can you verify that `ocnice_prep` is working on Ursa. All log files are here: `/scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/ocnice_prep`

GeorgeGayno-NOAA commented:
All `cpld_gridgen` tests failed on Ursa. (6bceac5). Differences from the (Hera) baseline appear to be small. Here are the some of the differences from the `nccmp_900.log`:

```
Variable  Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
land_frac /       516 -1.28378e-13 7.31674e-13 -7.77156e-15 5.66214e-15 1.34337e-14 -2.48795e-16 1.89019e-15
Variable Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
area_a   /       682 -3.25261e-17 7.67008e-15 -4.85723e-17 4.51028e-17 9.36751e-17 -4.76922e-20 1.47867e-17
area_b   /       748 -9.36751e-17  1.0502e-14 -9.36751e-17 7.97973e-17 1.73472e-16 -1.25234e-19 1.95688e-17
frac_a   /       596   9.4369e-15 3.92686e-13 -2.33147e-15 3.66374e-15  5.9952e-15  1.58337e-17 8.05017e-16
frac_b   /       729 -2.42475e-14 5.21633e-13 -3.55271e-15 2.88658e-15 6.43929e-15 -3.32614e-17 9.12855e-16
S        /      2689 -2.09361e-14 1.10651e-12 -3.21965e-15 3.77476e-15 6.99441e-15 -7.78584e-18 6.28342e-16
```

@DeniseWorthen - again I need your expertise. All log files are here: `/scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/cpld_gridgen`

DeniseWorthen commented:
@GeorgeGayno-NOAA I think the cpld_gridgen baseline differences are fine if you're comparing against the Hera-generated baselines. The compilers have been updated on Ursa relative to Hera, so we're seeing all baselines changing in UFS also. 

For the ocnice_prep, are they also comparing against Hera-generated baselines?

DeniseWorthen commented:
@GeorgeGayno-NOAA I can see there are large differences being generated along the equator (and one row south of the EQ) where the utility is filling the layer values beyond where the water column is "wet". It is very odd---I will need to dig into what is happening. I think this is some kind of artifact and the utility is generating a valid warmstart but I need to check.

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA I think the cpld_gridgen baseline differences are fine if you're comparing against the Hera-generated baselines. The compilers have been updated on Ursa relative to Hera, so we're seeing all baselines changing in UFS also.
> 
> For the ocnice_prep, are they also comparing against Hera-generated baselines?

Yes.

GeorgeGayno-NOAA commented:
The `regrid_sfc` test failed on Ursa (using 6bceac5):

```
CHECK sfci.tile1.nc
+ /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/nccmp-1.9.0.1-nuom2du/bin/nccmp -dmfqS sfci.tile1.nc /scratch3/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/regrid_sfc//baseline_data/gauss2fv3incr/sfci.tile1.nc
Variable   Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
soilt1_inc /     14674  4.22339e-13 2.29594e-11 -4.09672e-14 4.28546e-14 8.38218e-14  2.87815e-17 3.29287e-15
soilt2_inc /     14790  1.53172e-13 1.01955e-11 -1.87073e-14 4.37428e-14   6.245e-14  1.03564e-17 1.55117e-15
slc1_inc   /     14842 -1.48019e-14 3.24704e-13 -1.79804e-15 8.82107e-16 2.68015e-15 -9.97301e-19 6.22858e-17
slc2_inc   /     14881  -6.0238e-15 9.97753e-14 -4.63171e-16 4.41921e-16 9.05092e-16 -4.04798e-19 2.15929e-17
```

These differences (using the Hera baselines) appear to be small, but I will verify with @ClaraDraper-NOAA.

ClaraDraper-NOAA commented:
> The `regrid_sfc` test failed on Ursa (using [6bceac5](https://github.com/ufs-community/UFS_UTILS/commit/6bceac5863f9394d8954da2d6fbe3e06be658d81)):
> 
> ```
> CHECK sfci.tile1.nc
> + /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/nccmp-1.9.0.1-nuom2du/bin/nccmp -dmfqS sfci.tile1.nc /scratch3/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/regrid_sfc//baseline_data/gauss2fv3incr/sfci.tile1.nc
> Variable   Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
> soilt1_inc /     14674  4.22339e-13 2.29594e-11 -4.09672e-14 4.28546e-14 8.38218e-14  2.87815e-17 3.29287e-15
> soilt2_inc /     14790  1.53172e-13 1.01955e-11 -1.87073e-14 4.37428e-14   6.245e-14  1.03564e-17 1.55117e-15
> slc1_inc   /     14842 -1.48019e-14 3.24704e-13 -1.79804e-15 8.82107e-16 2.68015e-15 -9.97301e-19 6.22858e-17
> slc2_inc   /     14881  -6.0238e-15 9.97753e-14 -4.63171e-16 4.41921e-16 9.05092e-16 -4.04798e-19 2.15929e-17
> ```
> 
> These differences (using the Hera baselines) appear to be small, but I will verify with @ClaraDraper-NOAA.

Yes, those differences are very small.

GeorgeGayno-NOAA commented:
All `chgres_cube` regression tests failed (using 6bceac5). The Hera baselines were used.  Differences were not significant.   For example, tile1 differences from the C192 FV3 HISTORY test:

The tile 1 'atm' file:

```
CHECK out.atm.tile1.nc
+ nccmp -dmfqS out.atm.tile1.nc /scratch3/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/chgres_cube/baseline_data/c192_fv3_history/out.atm.tile1.nc
Variable Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
delp     /         1  0.000244141 0.000244141  0.000244141 0.000244141           0  0.000244141           0
liq_wat  /       604  2.18585e-15  4.0247e-15 -4.44089e-16 8.88178e-16 1.33227e-15  3.61896e-18 5.19484e-17
u_s      /         1  5.68434e-14 5.68434e-14  5.68434e-14 5.68434e-14           0  5.68434e-14           0
v_s      /         3 -1.86265e-09 6.14673e-08 -2.98023e-08 2.98023e-08 5.96046e-08 -6.20882e-10 2.98217e-08
```

The tile 1 'sfc' file:

```
CHECK out.sfc.tile1.nc
+ nccmp -dmfqS out.sfc.tile1.nc /scratch3/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/chgres_cube/baseline_data/c192_fv3_history/out.sfc.tile1.nc
Variable Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
f10m     /     24720  4.67404e-14 6.75027e-12 -8.88178e-15  5.9952e-15  1.4877e-14  1.89079e-18 4.77368e-16
t2m      /     15079 -3.80851e-12 9.15122e-10 -3.41061e-13 3.41061e-13 6.82121e-13  -2.5257e-16 6.26276e-14
q2m      /     27991 -2.29829e-15 1.70654e-13 -1.30104e-16 2.08167e-16 3.38271e-16 -8.21082e-20 1.00562e-17
uustar   /     33499   1.3582e-13 1.67405e-11 -2.08722e-14 2.56462e-14 4.65183e-14  4.05445e-18  1.0676e-15
ffmm     /     27869  7.58194e-12 3.13606e-10 -8.56204e-13 7.81597e-13  1.6378e-12  2.72056e-16 3.35133e-14
tprcp    /     20679 -7.51598e-17 7.14539e-15 -1.92988e-17 1.77267e-17 3.70255e-17 -3.63459e-21 9.98102e-19
```

GeorgeGayno-NOAA committed:
Fix typo in regression test driver script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Clean up regression test driver script.

Fixes #1064.

GeorgeGayno-NOAA committed:
Update comments in CMakelists.txt concerning use
of sp and ip v5+

Fixes #1064.

GeorgeGayno-NOAA committed:
Update to spack stack v1.9.2

Fixes #1068

GeorgeGayno-NOAA commented:
> All consistency tests were run on Ursa using the branch at [6bceac5](https://github.com/ufs-community/UFS_UTILS/commit/6bceac5863f9394d8954da2d6fbe3e06be658d81). The baselines were copied from Hera.
> 
> The following tests all passed:
> 
> * weight_gen
> * snow2mdl
> * ice_blend
> * global_cycle
> 
> The other tests had failures. Details will be provided.

These tests were rerun using spack stack 1.9.2 (c254bd8). All tests passed.

GeorgeGayno-NOAA commented:
> All `ocnice_prep` tests failed on Ursa ([6bceac5](https://github.com/ufs-community/UFS_UTILS/commit/6bceac5863f9394d8954da2d6fbe3e06be658d81)). Differences from the (Hera) baseline appear to be small. Here are the differences from the `nccmp_050_ocean.log` file:
> 
> ```
> Variable Group    Count         Sum    AbsSum         Min        Max      Range         Mean      StdDev
> sfc      /       219941 -0.00608705 0.0396815 -0.00443655 0.00203139 0.00646795 -2.76758e-08 1.65466e-05
> eta      /     11540193     -530683    616446    -1902.77    414.788    2317.56   -0.0459856     5.24544
> Temp     /     13415140    -1167.99   1793.17    -6.70048    2.81754    9.51802 -8.70649e-05   0.0212207
> Salt     /      8242225     98.0152   284.894   -0.440129    1.37529    1.81541  1.18918e-05    0.004976
> h        /     10253080     17246.3     20784    -252.123    411.662    663.786   0.00168206    0.586818
> u        /     19370198    -1.17202   51.7173  -0.0749373  0.0754253   0.150363 -6.05062e-08 0.000273376
> v        /     19330990     6.62261   46.9424   -0.037016  0.0818302   0.118846  3.42591e-07 0.000240951
> ```
> 
> @DeniseWorthen - can you verify that `ocnice_prep` is working on Ursa. All log files are here: `/scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/ocnice_prep`

These tests were rerun using spack stack 1.9.2 (c254bd8). The 1.9.1 files were used as the baseline. All tests using 1.9.2 failed, but the differences were smaller than the 1.9.1 tests, which compared against the Hera baselines. The `nccmp_050_ocean.log` file showed:

```
Variable Group    Count         Sum      AbsSum          Min         Max       Range        Mean      StdDev
u        /     20423804 4.53313e-11 6.01525e-09 -2.16493e-14 2.51466e-14 4.67959e-14 2.21953e-18   6.635e-16
v        /     20592284 1.08938e-10 1.52153e-08 -7.13318e-14 7.37188e-14 1.45051e-13 5.29026e-18 1.83274e-15
```

@DeniseWorthen - All 1.9.2 log files are here: `/scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/ocnice_prep/logs.192` if you want to check.

GeorgeGayno-NOAA commented:
> All seven `grid_gen` tests failed using [6bceac5](https://github.com/ufs-community/UFS_UTILS/commit/6bceac5863f9394d8954da2d6fbe3e06be658d81). In all tests, the 'orog' files were identical to the (Hera) baseline. Differences were noted in the 'grid' files and in a subset of the surface climatology files. These differences appeared to be small. And these differences were noted at a small percentage of the total model grid points (the test used a 222x204 grid). As an example, here are some differences from the GFDL regional test:
> 
> C424_grid.tile7.halo0.nc file:
> 
> ```
> CHECK C424_grid.tile7.halo0.nc
> + /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/nccmp-1.9.0.1-nuom2du/bin/nccmp -dmfqS C424_grid.tile7.halo0.nc /scratch3/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/grid_gen/baseline_data/gfdl.regional/C424_grid.tile7.halo0.nc
> Variable Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
> x        /        19 -1.08002e-12 1.19371e-12 -1.13687e-13 5.68434e-14  1.7053e-13 -5.68434e-14 3.28186e-14
> y        /       271 -4.08562e-13 2.69651e-12 -3.55271e-14 2.13163e-14 5.68434e-14 -1.50761e-15 1.19918e-14
> dx       /       273 -8.73115e-11 9.26084e-08 -4.70754e-09 4.58567e-09 9.29322e-09 -3.19822e-13 1.15624e-09
> dy       /       324 -3.81988e-11 3.08568e-07 -3.53793e-09 3.53793e-09 7.07587e-09 -1.17897e-13 1.20299e-09
> area     /       201    -0.108159     7.60721   -0.0721062   0.0721062    0.144212 -0.000538106   0.0387429
> ```
> 
> C424.vegetation_greenness.tile7.halo4.nc file:
> 
> ```
> CHECK ./sfc/C424.vegetation_greenness.tile7.halo4.nc
> + /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/nccmp-1.9.0.1-nuom2du/bin/nccmp -dmfqS ./sfc/C424.vegetation_greenness.tile7.halo4.nc /scratch3/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/grid_gen/baseline_data/gfdl.regional/./sfc/C424.vegetation_greenness.tile7.halo4.nc
> Variable             Group Count         Sum      AbsSum         Min         Max Range        Mean StdDev
> vegetation_greenness /         1 5.96046e-08 5.96046e-08 5.96046e-08 5.96046e-08     0 5.96046e-08      0
> ```
> 
> I don't think these differences indicate a problem running on Ursa.

The tests were rerun using spack stack 1.9.2 (c254bd8). The 1.9.1 results were used as the baseline. All tests passed.

GeorgeGayno-NOAA committed:
Add the load of doxygen to the build module.

Fixes #1068.

GeorgeGayno-NOAA commented:
Added ability to create Doxygen on Ursa (c42fa4d). However, it can't find the INPUT files. This is odd because I know the `chgres_cube.md` file does exist.

```
Searching for dia files...
Searching for files to exclude
Searching INPUT for files to process...
Searching for files in directory /scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/sorc/chgres_cube.fd
error: Specified markdown mainpage 'chgres_cube.md' does not exist
 (warning treated as error, aborting now)
```

The same build on Hera runs with no problems:

```
Searching for msc files...
Searching for dia files...
Searching for files to exclude
Searching INPUT for files to process...
Searching for files in directory /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.upstream/sorc/chgres_cube.fd
Reading and parsing tag files
Parsing files
Reading /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.upstream/sorc/chgres_cube.fd/docs/chgres_cube.md...
Preprocessing /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.upstream/sorc/chgres_cube.fd/atm_input_data.F90...
Parsing file /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.upstream/sorc/chgres_cube.fd/atm_input_data.F90...
....
```

Very odd.

BrianCurtis-NOAA commented:
What command are you running specifically? i'll try on my end

BrianCurtis-NOAA commented:
> What command are you running specifically? i'll try on my end

i ran:
```
mkdir build
cd build
cmake -DENABLE_DOCS=ON ..
make doc
```
and it seemed to run OK.

GeorgeGayno-NOAA commented:
> What command are you running specifically? i'll try on my end

I can run the part that is failing interactively as follows:

```
> cd /scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/build/sorc/chgres_cube.fd/docs
> doxygen Doxyfile
```

I tried hardwiring paths and copying the `chgres_cube.fd` to different directories. Nothing seems to work. The doxygen version on Ursa is newer than on Hera, I don't know if that is an issue.

BrianCurtis-NOAA commented:
I can reproduce it that way

GeorgeGayno-NOAA commented:
> seems to be an issue with the version of doxygen, i wonder if we could build a local doxygen version newer then the current and see if it still happens. Otherwise a workaround was made to work with other versions on other machines and those will need reverted for this version?

Thanks for verifying that it is not working. I thought I was going crazy. What type of workarounds were done for other machines? 

Let me ask the library's team if they have encountered this problem.

AlexanderRichert-NOAA commented:
I think we ran into this a long time ago with NCEPLIBS, maybe related to a version change in Doxygen (may be related to this issue: https://github.com/doxygen/doxygen/issues/10110). What worked for me just now with UFS_UTILS is changing
```
USE_MDFILE_AS_MAINPAGE = global_cycle.md
```
to
```
USE_MDFILE_AS_MAINPAGE = @abs_top_srcdir@/sorc/global_cycle.fd/docs/global_cycle.md
```
This appears to have built everything, though I also disabled WARN_AS_ERROR to side step failures related to undocumented routines (in at least one instance it's a Fortran interface that needs a dummy Doxygen entry).

If it's of any use, here are a couple other Doxygen versions I installed:
 - /contrib/spack-stack/spack-stack-1.9.2/envs/doxygen/install/gcc/12.4.0/doxygen-1.9.7-5vh3a45/bin/doxygen
 - /contrib/spack-stack/spack-stack-1.9.2/envs/doxygen/install/gcc/12.4.0/doxygen-1.12.0-32knhax/bin/doxygen

GeorgeGayno-NOAA commented:
If I try @AlexanderRichert-NOAA versions of doxygen:

```
diff --git a/modulefiles/build.ursa.intelllvm.lua b/modulefiles/build.ursa.intelllvm.lua
index bd0bbfa2..cff35b90 100644
--- a/modulefiles/build.ursa.intelllvm.lua
+++ b/modulefiles/build.ursa.intelllvm.lua
@@ -53,7 +53,9 @@ load(pathJoin("esmf", esmf_ver))
 nco_ver=os.getenv("nco_ver") or "5.2.4"
 load(pathJoin("nco", nco_ver))

-load("doxygen")
+prepend_path("MODULEPATH", "/contrib/spack-stack/spack-stack-1.9.2/envs/doxygen/install/modulefiles/gcc/12.4.0")
+load("doxygen/1.12.0")
```

I get the same error:

```
Doxygen version used: 1.12.0
Notice: Output directory '../../../docs/html/chgres_cube' does not exist. I have created it for you.
Generating docs for namespace dcstgeo
Searching for include files...
Searching for example files...
Searching for images...
Searching for dot files...
Searching for msc files...
Searching for dia files...
Searching for files to exclude
Searching INPUT for files to process...
Searching for files in directory /scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/sorc/chgres_cube.fd
Generating docs for namespace pesg
error: Specified markdown mainpage 'chgres_cube.md' does not exist
 (warning treated as error, aborting now)
Exiting...
```

Using 1.9.7 does not fix it either.

GeorgeGayno-NOAA committed:
Fix the Doxygen.in files so that is works with the Ursa version
of the doxygen executable.

Fixes #1051.

GeorgeGayno-NOAA commented:
@BrianCurtis-NOAA and @DavidHuber-NOAA - I think I have addressed your comments. The Doxygen is working, and I completed a few final tests. I would like to merge this week.

DavidHuber-NOAA reviewed (approved):
Looks good.  Thanks @GeorgeGayno-NOAA!

GeorgeGayno-NOAA committed:
Merge branch 'develop' into ursa_port

Fixes #1064.

GeorgeGayno-NOAA commented:
> After today's merge (#1061), the `cpld_gridgen` and `ocnice_prep` programs were updated and required new baselines. Rerun the regression tests on Ursa using [5e0310b](https://github.com/ufs-community/UFS_UTILS/commit/5e0310b450b3219108f4fd2b3c31d044ae0b07c0) and the new Hera baselines.
> 
> The `ocnice_prep` tests all failed. The log files are here:
> 
> * `/scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/ocnice_prep/logs.after.pr1061`
> 
> The differences from the baseline were similar to initial tests on Ursa before the #1061 merge. Those tests also used the Hea baseline and were deemed to be correct (See [#1068 (comment)](https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996385394)). The log files from that test are here:
> 
> * `/scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/ocnice_prep/logs.191`
> 
> @DeniseWorthen can you please check.

After the #1061 merge, the 0.25-degree `cpld_gridgen` baseline was updated (includes more tripole files). The new Hera baseline was moved to Ursa and the test was rerun. The log files are here:

- /scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/cpld_gridgen/logs.after.pr1061

These differences were similar to initial tests on Ursa before the #1061 merge (which were deemed to be correct. See: https://github.com/ufs-community/UFS_UTILS/pull/1068#issuecomment-2996546983) The log files from that test are here:

- /scratch3/NCEPDEV/da/George.Gayno/UFS_UTILS/reg_tests/cpld_gridgen/logs.191

@DeniseWorthen - this looks ok to me.

DeniseWorthen commented:
@BrianCurtis-NOAA Yes.
